### PR TITLE
feat: collect Drive logs with ELK

### PR DESCRIPTION
**Pull request by strophy**
An advanced log processing system is required to debug testnet. This PR deploys ELK services to the web node and Filebeat to masternodes to help monitor log files.

## Issue being fixed or feature implemented
Unable to debug non-determinism issues in js-drive on masternodes

## What was done?
- increased web instance storage and type
- added ELK services as docker-compose role in ansible
- modified security groups and load balancer
- added configuration for Kibana
   - [x] map `@timestamp` to `json.time`
   - [x] map `log.level` to `json.level`

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: terraform/aws/security_groups.tf [hunk 1]
```diff
@@ -175,6 +175,18 @@ resource "aws_security_group" "http" {
     ])
   }
 
+    ingress {
```
Comment (shumkov) [inline terraform/aws/security_groups.tf:178]: Formatting

File: ansible/roles/mn-elk-services/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  loop:
+    - docker-compose.yml
+    - elasticsearch.yml
+    - kibana.yml
+    - logstash.yml
+    - logstash.conf
+
+- name: start elk services
+  docker_compose:
+    project_src: '{{ mn_elk_services_path }}'
```
Comment (shumkov) [inline ansible/roles/mn-elk-services/tasks/main.yml:22]: elk_path

File: ansible/deploy.yml [hunk 3]
```diff
@@ -194,6 +194,7 @@
       dashd_zmq: true
       dashd_listen: true
     - insight
+    - mn-elk-services
```
Comment (shumkov) [inline ansible/deploy.yml:197]: I think elk is not related to mn logic, let's just call this role `elk`

File: ansible/group_vars/all [hunk 4]
```diff
@@ -117,3 +117,10 @@ swap_space: 2G
 
 mn_evo_services_compose_project_name: mn_evo_services
 mn_evo_services_path: '{{ dashd_home }}/{{ mn_evo_services_compose_project_name }}'
+
+# ELK
+
+mn_elk_services_compose_project_name: mn_elk_services
```
Comment (shumkov) [inline ansible/group_vars/all:123]: elk_compose_project_name: elk

File: ansible/group_vars/all [hunk 5]
```diff
@@ -117,3 +117,10 @@ swap_space: 2G
...
+mn_elk_services_path: '{{ dashd_home }}/{{ mn_elk_services_compose_project_name}}'
+elasticsearch_password: password
```
Comment (shumkov) [inline ansible/group_vars/all:125]: I would add `elk_` prefix to all variables related to elk role.

File: ansible/group_vars/all [hunk 6]
```diff
@@ -117,3 +117,10 @@ swap_space: 2G
 
 mn_evo_services_compose_project_name: mn_evo_services
 mn_evo_services_path: '{{ dashd_home }}/{{ mn_evo_services_compose_project_name }}'
+
+# ELK
```
Comment (shumkov) [inline ansible/group_vars/all:121]: I would recommend to move these variables to role (`elk/vars/main.yaml`)

File: ansible/roles/elastic.beats/.kitchen.yml [hunk 7]
```diff
@@ -0,0 +1,126 @@
+---
```
Comment (shumkov) [inline ansible/roles/elastic.beats/.kitchen.yml:1]: Why not just add this role to `requirements.yml` and use `ansible-galaxy install -r requirements.yml`?

Comment (strophy) [reply ansible/roles/elastic.beats/.kitchen.yml]: Is [this](https://github.com/dashevo/dash-network-deploy/pull/217/commits/2de94b846551bb81d1eaef3fb10fec8b38aa8455) what you mean?

File: ansible/deploy.yml [hunk 8]
```diff
@@ -111,6 +111,24 @@
...
         - evo
       when: evo_services
+    - include_role:
+      name: elastic.beats
+      vars:
+        beat: filebeat
+        beat_conf:
+          filebeat:
+            inputs:
+              - type: log
```
Comment (shumkov) [inline ansible/deploy.yml:121]: We need to enable json, set meesage_key, timestamp and other options

File: ansible/deploy.yml [hunk 9]
```diff
@@ -133,20 +153,26 @@
...
-      tags:
-        - dashd
-    - mn-sentinel
-    - role: insight
-      ui: false
-      when: evo_services
-    - role: mn-evo-services
-      tags:
-        - evo
-      when: evo_services
```
Comment (shumkov) [inline ansible/deploy.yml:149]: We still need this stuff

File: terraform/aws/variables.tf [hunk 10]
```diff
@@ -133,3 +138,8 @@ variable "mn_node_disk_size" {
   description = "Default disk size for nodes, increase for testnet"
   default     = 30
 }
+
+variable "web_node_disk_size" {
+  description = "Default disk size for nodes, increase for testnet"
```
Comment (shumkov) [inline terraform/aws/variables.tf:143]:
```suggestion
  description = "Disk size of web nodes"
```

File: terraform/aws/variables.tf [hunk 11]
```diff
@@ -133,3 +138,8 @@ variable "mn_node_disk_size" {
   description = "Default disk size for nodes, increase for testnet"
```
Comment (shumkov) [inline terraform/aws/variables.tf:138]: Disk size of masternodes

File: ansible/deploy.yml (hunk 9 above)
Comment (strophy) [reply ansible/deploy.yml]: Yeah, committed by accident after trying to run it

File: ansible/roles/mn-evo-services/tasks/main.yml [hunk 12]
```diff
@@ -22,13 +22,13 @@
       - name: platform-logs
         definitions:
           - logs:
-              - '{{ mn_evo_services_path }}/logs/*.log'
+              - '{{ mn_evo_services_path }}/logs/*.log*'
```
Comment (shumkov) [inline ansible/roles/mn-evo-services/tasks/main.yml:25]: Do we need the star at the end for logrotate?

Comment (strophy) [reply ansible/roles/mn-evo-services/tasks/main.yml:25]: I replaced it while trying to understand why the container wasn't releasing the file after it was renamed so we could still capture logs from files with a date suffix. The last two hosts exhibiting this behaviour (seed nodes) have been fixed now so I guess we can remove it.

File: ansible/deploy.yml [hunk 13]
```diff
@@ -111,6 +111,7 @@
       tags:
         - evo
       when: evo_services
+    - role: filebeat
```
Comment (shumkov) [inline ansible/deploy.yml:114]: let's add an `elk-` prefix to group logically together with `elk-stack`?

File: ansible/roles/filebeat/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  vars:
+    beat: filebeat
+    beat_conf:
+      filebeat:
+        inputs:
+          - type: log
+            enabled: true
+            paths:
+              - "{{ mn_evo_services_path }}/logs/*.log*"
+            json.keys_under_root: false
```
Comment (shumkov) [inline ansible/roles/filebeat/tasks/main.yml:15]: It's `false` by default.

File: ansible/group_vars/all [hunk 15]
```diff
@@ -124,3 +124,4 @@ mn_evo_services_path: '{{ dashd_home }}/{{ mn_evo_services_compose_project_name
 
 elk_elasticsearch_username: elastic
 elk_elasticsearch_password: password
+elk_kibana_encryptionkey: 32_chars_or_longer_encryption_key
```
Comment (shumkov) [inline ansible/group_vars/all:127]: Let's keep it empty so if you forgot to set it it will raise an error

File: terraform/aws/variables.tf [hunk 16]
```diff
@@ -130,6 +135,11 @@ variable "core_node_disk_size" {
...
 
 variable "mn_node_disk_size" {
-  description = "Default disk size for nodes, increase for testnet"
+  description = "Disk size of masternodes"
   default     = 30
 }
+
+variable "web_node_disk_size" {
+  description = "Disk size of web nodes"
+  default     = 200
```
Comment (shumkov) [inline terraform/aws/variables.tf:144]: Let's set 100

File: ansible/roles/filebeat/templates/ilm_policy.json.j2 [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+          },
+          "migrate": {
+            "enabled": false
+          }
+        }
+      },
+      "cold": {
+        "min_age": "14d",
+        "actions": {
+          "freeze": {},
```
Comment (shumkov) [inline ansible/roles/filebeat/templates/ilm_policy.json.j2:27]: We don't need to pass anything here like `"enabled": true`?

File: ansible/roles/dashd/templates/dash.conf.j2 [hunk 18]
```diff
@@ -14,6 +14,8 @@ maxconnections=256
 debug={{ dashd_debug }}
 printtoconsole=1
 
+llmq-qvvec-sync=llmq_100_67
```
Comment (shumkov) [inline ansible/roles/dashd/templates/dash.conf.j2:17]: This one is to enable quorum public key shares.. Let's extract it to separate PR and we need to update MN boostrap too.

File: ansible/roles/filebeat/templates/ilm_policy.json.j2 (hunk 17 above)
Comment (strophy) [reply ansible/roles/filebeat/templates/ilm_policy.json.j2:27]: No, it doesn't seem like it. `auto` is the default if `setup.ilm.enabled` is not defined, and the ILM has been successfully created and enabled already using this config.

Comment (shumkov) [reply ansible/roles/filebeat/templates/ilm_policy.json.j2:27]: What `auto` means? Will it freeze indices on cold phaze?

**Comment by shumkov** [comment]
@strophy Also:
- wildcards don't work with `json.msg` (`json.msg: Block commit*`)
- make `json.msg` as a default field for index

File: ansible/roles/filebeat/templates/ilm_policy.json.j2 (hunk 17 above)
Comment (strophy) [reply ansible/roles/filebeat/templates/ilm_policy.json.j2:27]: Yes, freeze is configured. You can check the config created by this policy file in Kibana under Stack Management > Index Lifecycle Policies > Filebeat.

But freezing will not reduce the size on disk, it only reduces RAM use...

File: ansible/roles/filebeat/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,136 @@
+---
+
+- name: copy filebeat ILM policy
+  template:
+    src: 'ilm_policy.json.j2'
```
Comment (shumkov) [inline ansible/roles/filebeat/tasks/main.yml:5]: If you do not use templates you can put to `files` fir instead of `templates` and use https://docs.ansible.com/ansible/2.5/modules/copy_module.html

File: ansible/deploy.yml (hunk 13 above)
Comment (strophy) [reply ansible/deploy.yml:114]: OK, I looked at how elastic name their products and they strongly discourage use of `ELK` now. Everything is called Elastic Stack, and Filebeat is part of Elastic Beats (which include many different types of collectors). So I have named it `elastic-stack` and `elastic-beats` to be consistent with their naming.

Comment (shumkov) [reply ansible/deploy.yml:114]: Sounds good!

File: ansible/roles/elastic-beats/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+              - json.hostname
+              - json.time
+            ignore_missing: true
+    output_conf:
+      elasticsearch:
+        hosts:
+          - "{{ hostvars['web-1'].private_ip }}:9200"
+        username: "{{ elastic_username }}"
+        password: "{{ elastic_password }}"
+        index: "logs-drive-abci-%{[agent.version]}"
```
Comment (shumkov) [inline ansible/roles/elastic-beats/tasks/main.yml:136]: We can move `index` under `input`

File: ansible/roles/elastic-beats/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+      filebeat:
+        inputs:
+          - type: log
+            enabled: true
+            json.message_key: msg
+            json.keys_under_root: false
+            exclude_files: ['\.gz$']
+            paths:
+              - "{{ mn_evo_services_path }}/logs/*.log*"
+      processors:
```
Comment (shumkov) [inline ansible/roles/elastic-beats/tasks/main.yml:61]: We can move processors under `input`: https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-log.html#_processors_10

File: ansible/roles/elastic-beats/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+
+- name: set up filebeat log monitoring
+  include_role:
+    name: elastic.beats
+  vars:
+    beat: filebeat
+    beat_conf:
+      setup:
+        template:
+          name: logs-drive-abci
```
Comment (shumkov) [inline ansible/roles/elastic-beats/tasks/main.yml:19]: We need to add version here, like "logs-drive-abci-%{[agent.version]}"

File: ansible/roles/elastic-beats/tasks/main.yml [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+          pattern: logs-drive-abci-*
```
Comment (shumkov) [inline ansible/roles/elastic-beats/tasks/main.yml:20]: "logs-drive-abci-%{[agent.version]}-*"

File: ansible/roles/elastic-beats/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+                field:
+                  - json.height
+                  - json.abciMethod
+                  - json.level
+                order:
+                  - asc
+                  - asc
+                  - asc
+        ilm:
+          policy_file: /etc/filebeat/ilm_policy.json
```
Comment (shumkov) [inline ansible/roles/elastic-beats/tasks/main.yml:51]: we need also:
```
rollover_alias: "logs-drive-abci"
policy_name: "masternode-logs"
```

**Comment by shumkov** [comment]
There is a system field in an index document called `default` that combines values from other fields. It allows you to search through documents without specifying fields. We will reduce index size a lot if we keep the only `message` there or even disable it at all (I guess it's better).

I found a way how to keep the only the message there. We need to update this file before we generate default fileds.yml config: https://github.com/elastic/ecs/pull/687/files#diff-3e8942b604283b26a1b999f8effd915bc4e3b6eec68301824df502457e897debR223

Maybe you will find a way how to disable the default field at all.

File: ansible/deploy.yml [hunk 25]
```diff
@@ -194,6 +197,7 @@
       dashd_zmq: true
       dashd_listen: true
     - insight
+    - elk
```
Comment (shumkov) [inline ansible/deploy.yml:200]: it should be `elastic-stack`

File: ansible/deploy.yml [hunk 26]
```diff
@@ -147,6 +148,8 @@
       tags:
         - evo
       when: evo_services
+    - role: filebeat
```
Comment (shumkov) [inline ansible/deploy.yml:151]: it should be `elastic-filebeat`

File: ansible/deploy.yml (hunk 13 above)
Comment (shumkov) [reply ansible/deploy.yml:114]: it should be `elastic-filebeat`

**Comment by strophy** [comment]
Closed in favour of #225
